Add IPv6 string validation

NeQuissimus opened the pull request:
Looks like IPv6 validation using a regex is what makes the most sense.

- I looked at `sun.net.util.IPAddressUtil` but that's an internal JDK API, not supposed to be used, not available for Scala.JS (it is the underlying implementation for `java.net.InetAddress`)
- `java.net.InetAddress` itself always tries to make DNS calls unless you convert the String into the address bytes. That would, however, require a lot of logic to undo compacting etc.
- Guava and Apache Commons Validator are JVM-only and fairly large

fthomas commented on modules/core/shared/src/main/scala/eu/timepit/refined/string.scala:
```diff
@@ -15,6 +15,14 @@ object string extends StringValidate with StringInference {
+  final case class IPv6()
+  object IPv6 {
+    val regex =
+      "^(^(([0-9A-Fa-f]{1,4}(((:[0-9A-Fa-f]{1,4}){5}::[0-9A-Fa-f]{1,4})|((:[0-9A-Fa-f]{1,4}){4}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,1})|((:[0-9A-Fa-f]{1,4}){3}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,2})|((:[0-9A-Fa-f]{1,4}){2}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,3})|(:[0-9A-Fa-f]{1,4}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,4})|(::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,5})|(:[0-9A-Fa-f]{1,4}) ...
```
Can we factor out common subexpressions to reduce the length of that regex? `[0-9A-Fa-f]{1,4}` seems to be a good candidate on a first glance. I'd imagine doing it like this:
```scala
val regex = {
  val group = "[0-9A-Fa-f]{1,4}"
  s"^(^(($group(((:$group){5}::$group) ..."
}
```

derekmorr commented:
java.net.InetAddress only makes DNS calls if the input isn't an IP address literal. IPv6 address parsing is pretty complicated. It doesn't look like this parser handles embedded IPv4-literals:

```scala

scala>     val regex =
     |       "^(^(([0-9A-Fa-f]{1,4}(((:[0-9A-Fa-f]{1,4}){5}::[0-9A-Fa-f]{1,4})|((:[0-9A-Fa-f]{1,4}){4}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,1})|((:[0-9A-Fa-f]{1,4}){3}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,2})|((:[0-9A-Fa-f]{1,4}){2}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,3})|(:[0-9A-Fa-f]{1,4}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,4})|(::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,5})|(:[0-9A-Fa-f]{1,4}){7}))$|^(::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,6})$)|^::$)|^((([0-9A-Fa-f]{1,4}(((:[0-9A-Fa-f]{1,4}){3}::([0-9A-Fa-f]{1,4}){1})|((:[0-9A-Fa-f]{1,4}){2}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,1})|((:[0-9A-Fa-f]{1,4}){1}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,2})|(::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,3})|((:[0-9A-Fa-f]{1,4}){0,5})))|([:]{2}[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,4})):|::)((25[0-5]|2[0-4][0-9]|[0-1]?[0-9]{0,2})\\.){3}(25[0-5]|2[0-4][0-9]|[0-1]?[0-9]{0,2})$$".r.pattern
regex: java.util.regex.Pattern = ^(^(([0-9A-Fa-f]{1,4}(((:[0-9A-Fa-f]{1,4}){5}::[0-9A-Fa-f]{1,4})|((:[0-9A-Fa-f]{1,4}){4}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,1})|((:[0-9A-Fa-f]{1,4}){3}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,2})|((:[0-9A-Fa-f]{1,4}){2}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,3})|(:[0-9A-Fa-f]{1,4}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,4})|(::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,5})|(:[0-9A-Fa-f]{1,4}){7}))$|^(::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,6})$)|^::$)|^((([0-9A-Fa-f]{1,4}(((:[0-9A-Fa-f]{1,4}){3}::([0-9A-Fa-f]{1,4}){1})|((:[0-9A-Fa-f]{1,4}){2}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,1})|((:[0-9A-Fa-f]{1,4}){1}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,2})|(::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,3})|((:[0-9A-Fa-f]{1,4}){0,5})))|([:]{2}[0-9A-Fa-f]{1,4}(:[0-9A-Fa-...

scala> regex.matcher("2001:db8:122:344::192.0.2.33").matches // example from https://tools.ietf.org/html/rfc6052#section-2.4
res1: Boolean = false

scala> import java.net.InetAddress // But the JDK's parser works fine
import java.net.InetAddress

scala> InetAddress.getByName("2001:db8:122:344::192.0.2.33")
res2: java.net.InetAddress = /2001:db8:122:344:0:0:c000:221
```

fthomas commented:
Thanks @derekmorr for this test case. https://stackoverflow.com/a/17871737/460387 shows that a regex can handle embedded IPv4 literals. @NeQuissimus what do you think about updating the regex to also support embedded IPv4 literals?

NeQuissimus commented:
Sorry, I completely missed the feedback here. I will take a look.
I was aware of this, my regex only validates against RFC-8200.
I'll see if I can't make some improvements or maybe find a better regex online.

kusamakura commented:
It might also be a nice touch to have an `Inference` for `IPv4 ==> IPv6`, if that's possible. (Or also the other way around for valid values.)

NeQuissimus commented:
I fixed up the regex to account for embedded IPv4 addresses and I also extracted common elements.
I added tests for the new address formats.

@fthomas Can you kick off the Travis build again? No matter how often I run the tests locally, they pass...

fthomas commented:
The Scala.js tests failed with
```
[info] ! StringValidate.IPv6.isValid.compact: Falsified after 0 passed tests.
```
This should be reproducible with `coreJS/test`.

NeQuissimus commented:
wow, this took me a while! JavaScript's Regex engine has different eagerness rules than the one in the JVM...
I fixed it now and hopefully Travis will agree...
